### PR TITLE
Added safelisting to the process tree creation flow

### PR DESCRIPTION
**Pull request by vipj**
Added all functions required to remove safelisted branches from process trees when they are created

### CODE REVIEW

File: assemblyline_v4_service/common/dynamic_service_helper.py [hunk 1]
```diff
@@ -408,9 +412,151 @@ def get_process_tree_with_signatures(self, signatures: List[Dict[str, Any]] = No
...
+                for node_hash in hashes:
+                    final_hash += node_hash
+
+                encoded_final_hash = final_hash.encode()
+                process_tree_hashes.append(sha256(encoded_final_hash).hexdigest())
+
+        return process_tree_hashes
+
+    @staticmethod
+    def _remove_safe_roots(process_tree: List[Dict[str, Any]], process_tree_hashes: List[str], safe_process_tree_hashes: Dict[str, Dict[str, Any]]) -> None:
```
Comment (cccs-kevin) [inline assemblyline_v4_service/common/dynamic_service_helper.py:483]: this method is no longer used, please remove with associated test method

**Review by cccs-kevin — CHANGES_REQUESTED** [review]

File: assemblyline_v4_service/common/dynamic_service_helper.py [hunk 2]
```diff
@@ -545,17 +530,21 @@ def _remove_safe_leaves(process_tree: List[Dict[str, Any]], leaf_hashes: List[Li
...
+
+        Returns:
+            List[Dict[str, Any]]: A list of processes in a tree structure, with the safe branches filtered out
+        """        
         leaf_hashes = SandboxOntology._create_hashes(process_tree, 'leaf')
         SandboxOntology._remove_safe_leaves(process_tree, leaf_hashes, safe_leaf_hashes)
 
         return process_tree
 
+
```
Comment (cccs-kevin) [inline assemblyline_v4_service/common/dynamic_service_helper.py:547]: this extra new line is not compliant with PEP 8 stylings

File: assemblyline_v4_service/common/dynamic_service_helper.py [hunk 3]
```diff
@@ -44,7 +45,7 @@ def _regex_substitution(path: str, rule: Dict[str, str]) -> str:
         rule['regex'] = rule['regex'].split('*')
         rule['regex'] = [escape(e) for e in rule['regex']]
         rule['regex'] = '[^\\\\]+'.join(rule['regex'])
-        path = sub(rf"{rule['regex']}", rule['replacement'], path)
+        path = sub(rf"{rule['regex']}", rule['replacement'], path)
```
Comment (cccs-kevin) [inline assemblyline_v4_service/common/dynamic_service_helper.py:48]: unnecessary space

**Review by cccs-kevin — CHANGES_REQUESTED** [review]

File: assemblyline_v4_service/common/dynamic_service_helper.py [hunk 4]
```diff
@@ -398,19 +399,140 @@ def _match_signatures_to_process_events(self, signature_dicts: List[Dict[str, An
...
+    
+    @staticmethod
+    def _create_hashed_node(parent: str, node: Dict[str, Any], hashes: List[str]) -> None:
+        """
+        This method takes a single node and hashes node attributes.
+        Recurses through children to do the same.
+        :param parent: A string representing the node hash
+        :param node: A dictionary representing the node
+        :param hashes: A list containing the hashes of the root
+        :param hash_type: A string representing what type of processing we are doing (root vs leaf)
```
Comment (cccs-kevin) [inline assemblyline_v4_service/common/dynamic_service_helper.py:426]: cut

File: assemblyline_v4_service/common/dynamic_service_helper.py [hunk 5]
```diff
@@ -398,19 +399,140 @@ def _match_signatures_to_process_events(self, signature_dicts: List[Dict[str, An
...
+            SandboxOntology._create_hashed_node(encoded256, child, hashes)
+
+    @staticmethod
+    def _create_hashes(process_tree: List[Dict[str, Any]]) -> List[List[str]]:
+        """
+        This method creates hashes for each root, or all root-to-leaf paths, in the tree, depending 
+        on which hash_type is specified
+        :param process_tree: A list of dictionaries where each dictionary represents a root.
+        :param hash_type: A string representing what type of processing we are doing (root vs leaf)
+        :return: A list of strings representing hashes, in the same order as the provided process tree
```
Comment (cccs-kevin) [inline assemblyline_v4_service/common/dynamic_service_helper.py:454]: update

File: assemblyline_v4_service/common/dynamic_service_helper.py [hunk 6]
```diff
@@ -398,19 +399,140 @@ def _match_signatures_to_process_events(self, signature_dicts: List[Dict[str, An
...
+                return None
+        
+    @staticmethod
+    def _remove_safe_leaves(process_tree: List[Dict[str, Any]], leaf_hashes: List[List[str]], safe_leaf_hashes: List[str]) -> None:
+        """
+        This method checks each leaf's hash against the safe hashes and removes safe branches from the process tree
+
+        Args:
+            process_tree (List[Dict[str, Any]]): A list of dictionaries where each dictionary represents a root.
+            leaf_hashes (List[str]): A list of strings representing hashes of each branch in the process tree
```
Comment (cccs-kevin) [inline assemblyline_v4_service/common/dynamic_service_helper.py:508]: wrong type and def

File: assemblyline_v4_service/common/dynamic_service_helper.py [hunk 7]
```diff
@@ -398,19 +399,140 @@ def _match_signatures_to_process_events(self, signature_dicts: List[Dict[str, An
...
+            # List to hold each hash computed using the _create_hashed_node function
+            hashes = []
+            SandboxOntology._create_hashed_node("", root, hashes)
+            
+            process_tree_hashes.append(hashes)
+
+        return process_tree_hashes
+
+    @staticmethod
+    def _remove_safe_leaves_helper(node: Dict[str, Any], leaf_hashes: List[str], safe_leaf_hashes: List[str]) -> Union[str, None]:
```
Comment (cccs-kevin) [inline assemblyline_v4_service/common/dynamic_service_helper.py:471]: leaf_hashes is wrong type

File: assemblyline_v4_service/common/dynamic_service_helper.py [hunk 8]
```diff
@@ -398,19 +399,140 @@ def _match_signatures_to_process_events(self, signature_dicts: List[Dict[str, An
...
+        """
+        This method is used to recursively remove safe branches from the given node. It removes a branch from the leaf 
+        up until it is no longer safelisted
+
+        Args:
+            node (Dict[str, Any]): A dictionary of a process tree node (root)
+            leaf_hashes (List[str]): The hashes of all of the leaves in the tree
```
Comment (cccs-kevin) [inline assemblyline_v4_service/common/dynamic_service_helper.py:478]: update_type and def

**Review by cccs-kevin — CHANGES_REQUESTED** [review]

**Review by cccs-kevin — APPROVED** [review]
TODO: make tests less text-heavy

**Comment by cccs-kevin** [comment]
Closes https://cccs.atlassian.net/browse/AL-1336
Incoming PR https://github.com/CybercentreCanada/assemblyline-service-cuckoo/pull/137 requires this to be merged first.

**Comment by cccs-kevin** [comment]
Note that tests passed in dev merge. Pipeline issue is due to this PR being made before the dev PR

**Review by cccs-rs — APPROVED** [review]

**Pull request merged**